### DISCUSSION TITLE
[Question][Hw3] solver does not terminate

### DISCUSSION THREAD

**Discussion by frogfrog2** [prosyslab-classroom/cs424-program-reasoning#281] [Q&A]
Name: Jihwan Park

My implementation works well for example 1,2,4,5, but I'm having trouble dealing with example 3.
Z3 solver does not terminate when I run the verifier with example3.ll file.
I couldn't find the exact reason yet, but I wonder if the following rule could be one of the reasons solver does not terminate.

`(forall ((%call Int)
         (%cmp Bool)
         (%conv Int)
         (%call1 Int)
         (%cmp2 Bool)
         (%conv3 Int)
         (%dec Int)
         (%inc Int)
         (%j.0 Int)
         (%k.0 Int))
  (=> (and entry
           (= %cmp (> %call 0))
           (>= %conv (ite %cmp 1 0))
           (<= %conv (ite %cmp 1 0))
           (> %conv 0)
           (= %cmp2 (> %call1 %call))
           (>= %conv3 (ite %cmp2 1 0))
           (<= %conv3 (ite %cmp2 1 0))
           (> %conv3 0))
      (while.cond %dec %inc 0 %call1 %conv3 %cmp2 %call1 %conv %cmp %call)))`

I found that the above rule is a little bit weird. As while.cond block has 10 free variables, I had to put %dec and %inc in the rule, which are not yet defined. To do so, I had to put %dec, %inc inside the quantifier 'forall'. However, this makes while.cond reachable with any integer values of %dec and %inc. I'm very confused now.

Is the above rule wrong??

**Comment by GetRusty** [comment]
In my opinion the rule is valid. Yes we need to put %dec and %inc in the rule of entry->while.cond, because while.cond has the free variables %dec and %inc.

The phrase "this makes while.cond reachable with any integer values of %dec and %inc." that you said is also true in my belief. Overflowing(not needed) quantifiers won't cause any problem, since it does not affect the other rules at all.

More important (in the implementation, at least) is about how we maintain a certain order of the quantifiers throughout the rules so the intended order of the parameters of a block is preserved. In the particular case of while..cond block in ex3, the third and fourth parameter should be set to the values of %k.0 and %j.0 always. But I think this is not related with the termination of a solver, so your problem may be based on other kinds of behaviors.

**Comment by frogfrog2** [comment]
Actually, my implementation substitutes 0 for %k.0, and %call1 for %j.0 using first two phi instructions of while.cond block.

**Comment by GetRusty** [comment]
> Actually, my implementation substitutes 0 for %k.0, and %call1 for %j.0 using first two phi instructions of while.cond block.

Oh I saw that, what I was going to say was your smt2 file might(and will) contain other logical flaws other than the number of parameters of each rules.
